### ISSUE TITLE
Angular 9.x Polyfills - Lighthouse Error - Does not use passive listeners to improve scrolling performance

### ISSUE THREAD

**Issue by naveedahmed1** [angular/angular#32733]
# 🐞 bug report

### Affected Package
Angular 9.x pollyfills

### Is this a regression?

I am not sure, I previously had score of 100 on Lighthouse in Chrome Dev tools, but recently it started displaying this error when I run audit for the website. It could a regression or may be the new version of Lighthouse just started reporting it.

### Description

When I run Audit for my website in Lighthouse in Chrome Dev tools, it displays below error:

`Does not use passive listeners to improve scrolling performance`

and points to pollyfills file (`/polyfills-es2015.5715214….js(www.1eent.com) | line: 0`)

## 🔬 Minimal Reproduction
Open chrome Dev Tools and in Audit tab, run audit for the website https://www.mustakbil.com
once the audit is complete, you can view the error under Best Practices section.

## 🔥 Exception or Error
<pre><code>
Does not use passive listeners to improve scrolling performance

/polyfills-es2015.5715214….js(www.1eent.com) | line: 0

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-next.4
Node: 10.16.3
OS: win32 x64
Angular: 9.0.0-next.6
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router, service-worker

Package                                    Version
--------------------------------------------------------------------
@angular-devkit/architect                  0.900.0-next.4
@angular-devkit/build-angular              0.900.0-next.4
@angular-devkit/build-optimizer            0.900.0-next.4
@angular-devkit/build-webpack              0.900.0-next.4
@angular-devkit/core                       9.0.0-next.4
@angular-devkit/schematics                 9.0.0-next.4
@angular/cdk                               8.1.4
@angular/cli                               9.0.0-next.4
@angular/fire                              5.2.1
@angular/material                          8.1.4
@angular/pwa                               0.900.0-next.4
@ngtools/webpack                           9.0.0-next.4
@nguniversal/module-map-ngfactory-loader   8.1.1
@schematics/angular                        9.0.0-next.4
@schematics/update                         0.900.0-next.4
rxjs                                       6.5.2
typescript                                 3.5.3
webpack                                    4.39.3
</code></pre>

Details about passive listeners can be found here https://developers.google.com/web/tools/lighthouse/audits/passive-event-listeners

**Comment by gkalpak** [comment]
Thx for reporting, @naveedahmed1. However, I wasn't able to reproduce the error by running a Lighthouse audit on https://www.mustakbil.com./ on Chrome 76.0.3809.132 (I get a score of 100 in "Best practices").

How are these listeners added.

(BTW, the problem with passive listeners is that they are currently not supported in all browsers that Angular supports and using them requires some [feature detection][1].)

[1]: https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md#feature-detection

**Comment by naveedahmed1** [comment]
@gkalpak can you please try it in Chrome Version 77?

For the listeners I add in my app, I specify `{passive: true}` as suggested here https://developers.google.com/web/tools/lighthouse/audits/passive-event-listeners?utm_source=lighthouse&utm_medium=devtools

But it seems that the issue is with the polyfills files, as its pointing to below files:
`/polyfills-es2015.8c0ceb8….js`


> (BTW, the problem with passive listeners is that they are currently not supported in all browsers that Angular supports and using them requires some feature detection.)

Yes I understand that, but since I am using Chrome and passive listeners are supported in Chrome, so it shouldn't throw this error in Chrome. I believe the listeners in polyfills has some issue.

**Comment by gkalpak** [comment]
OK, I see that now. It is hard to tell where the listener is coming from though. Could you try it on a non-minified version and try to find out where is the offending `addEventListener()` called?
(Happy to look into it myself as well, if you deploy a non-minified version somewhere :smiley:)

**Comment by naveedahmed1** [comment]
I tried to check it myself and it seems that there are multiple calls to nativeAddEventListener.call which could be the possible reason.

I have emailed you the link to the non-minified version.

**Comment by gkalpak** [comment]
Thx, @naveedahmed1. I see the error in the report, but I cannot understand why it happens.

Lighthouse reports that the violation happens in `polyfills-es2015.js` on line 1747. (Assuming lines are 0-indexed) on the 1748th line, there is indeed a call to `addEventListener()`:

```js
return nativeAddEventListener.call(taskData.target, taskData.eventName, taskData.capture ? globalZoneAwareCaptureCallback : globalZoneAwareCallback, taskData.options);
```

Based on the info the [Lighthouse audit docs][1], I put a conditional breakpoint on line 1748 like this:

```js
/^(wheel|mousewheel|touchstart|touch)$/i.test(taskData.eventName) && (typeof taskData.options !== 'object' || !taskData.options.passive)
```

The breakpoint was never hit 😕
AFAICT, this means that `addEventListener()` is never called with events `wheel`, `mousewheel`, `touchstart` or `touch` without `{passive: true}`. So, I do not understand why Lighthouse is complaining 😞

Maybe you could look on the Lighthouse issue tracker (or open an issue there)?

[1]: https://developers.google.com/web/tools/lighthouse/audits/passive-event-listeners

**Comment by patrickhulce** [comment]
@gkalpak this error in Lighthouse comes directly from Chrome. You can see the same violation being reported just in the DevTools console. Using a conditional breakpoint I did find at least two calls with `undefined` options instead of `{passive: false}` (see screenshot)

![image](https://user-images.githubusercontent.com/2301202/65848045-e2281800-e309-11e9-8954-5eb070ef83f9.png)

**Comment by gkalpak** [comment]
I don't see any such message in my DevTools 😞

**Comment by patrickhulce** [comment]
@gkalpak Violations are "Verbose" log level so they might be hidden for you. You can expand the console tab and filter just to verbose messages to make it easier.

![image](https://user-images.githubusercontent.com/2301202/65889497-f4846f00-e366-11e9-9237-5e5b08375a80.png)

**Comment by naveedahmed1** [comment]
Thank you so much @patrickhulce for sharing these details. @gkalpak I have below code in `main.ts` to avoid flickering issue that occurs when using Angular Universal:

```
document.addEventListener('DOMContentLoaded', () => {
  platformBrowserDynamic(providers).bootstrapModule(AppBrowserModule)
    .catch(err => console.log(err));
}, { passive: true });
```

As you can see that I have already added `{ passive: true }`, I don't know why is it still throwing error for this.

Can this be related to `zone.js`?

**Comment by gkalpak** [comment]
@patrickhulce: Yes, I have expanded the filter to show all messages (including verbose level ones) and I still don't see anything 🤷‍♂
I do see other "violations" fwiw:

```
[Violation] 'DOMContentLoaded' handler took 310ms
[Violation] Forced reflow while executing JavaScript took 84ms
[Violation] 'load' handler took 157ms
[Intervention] An <img> element was lazyloaded with loading=lazy, but had no dimensions specified. Specifying dimensions improves performance. See https://crbug.com/954323
```

##
@naveedahmed1: As mentioned above, only `wheel`, `mousewheel`, `touchstart` and `touch` events need to have `passive: true`, so adding it to other events is irrelevant.

**Comment by patrickhulce** [comment]
@gkalpak you're loading the mobile version of the page, correct? This Violation only seems to be present with mobile emulation. I realize that's implicit with a Lighthouse audit but might not be familiar to others :)

Also FWIW @naveedahmed1 this does appear to be flaky. I'm not able to reproduce in consistently reproduce in Canary Chrome DevTools for example, just stable Chrome.

**Comment by gkalpak** [comment]
I tried again with the mobile version and I got the warning once (but not on subsequent tries). It was indicating some `rocket-loader.min.js` file as the origin.

Sorry, I can't spend more time on this without a clear reproduction path.
If you can reproduce it int your environment and track the call back to its origin (and it turns out to be Angular), I would be happy to look into it further.

##
I am going to close the issue for now, since it is not actionable for us in its current form, but feel free to continue the discussion below and please _do_ open a new issue (referencing this one for context) once you have enough info (or a usable reproduction).

**Comment by naveedahmed1** [comment]
Thank you @gkalpak for your time let me investigate it further on my side.

@patrickhulce if I am not wrong, all event listeners attached to a page are also listed in the 'developers console', please take a look at the attached screenshots:

**All Event Listeners:**
![all-eventlistners](https://user-images.githubusercontent.com/7761180/65906447-09e7b080-e3dc-11e9-9a22-ce08b2957a1a.jpg)

**Blocking Event Listeners**
![Blocking-eventlistners](https://user-images.githubusercontent.com/7761180/65906505-2552bb80-e3dc-11e9-8975-0481f0440529.jpg)

As you can see that blocking event listeners doesn't include `wheel`, `mousewheel`, `touchstart` or `touch`.

Can you please also confirm if this check of event listeners was also present on the Lighthouse version available in version of Chrome prior to Chrome version 77? If so, then let me share that this wasn't reporting error on Chrome Version < 77.

**Comment by patrickhulce** [comment]
That's correct @naveedahmed1 it lists the currently active listeners in that pane.

I'm not able to see the non-passive touchstart listener in that panel even when I see the violation, so I'm guessing it's removed at some point. This is a good point on possible adjustments to make in LH, so we'll re-open that issue. Ideally, it's still never added in the first place, but it's not a cause for concern to the audit's purpose :)

Thanks for investigating all!

**Comment by naveedahmed1** [comment]
Thank you for the update @patrickhulce :)

> Can you please also confirm if this check of event listeners was also present on the Lighthouse version available in version of Chrome prior to Chrome version 77? If so, then let me share that this wasn't reporting error on Chrome Version < 77.

How about this? Since LH wasn't throwing this error in `Chrome 76.x`; @gkalpak also confirmed this in comment here https://github.com/angular/angular/issues/32733#issuecomment-534225445

**Comment by patrickhulce** [comment]
@naveedahmed1 I actually am able to reproduce this in Chrome 76.0.3809.100, I believe it's a combination of many different factors when it happens which has contributed to the trickiness in debugging.

**Comment by naveedahmed1** [comment]
I guess I have finally identified the issue. I think that the `evenListeners` that `matTooltip` add are not passive.

https://github.com/angular/components/blob/master/src/material/tooltip/tooltip.ts#L241

@gkalpak can you please re-check and confirm on your side?

I updated the conditional breakpoint to:

`/^(wheel|mousewheel|touchstart|touch)$/i.test(taskData.eventName) && taskData.options==undefined`

**Comment by naveedahmed1** [comment]
This issue was related to matTooltip which has been fixed in https://github.com/angular/components/pull/17003.

Thank you so much @gkalpak and @patrickhulce for your time and help :)

**Comment by gkalpak** [comment]
Awesome! Thx for the update, @naveedahmed1 :+1:
